clair/clair.go

```diff
@@ -89,29 +92,48 @@ func newLayer(image *docker.Image, index int) *layer {
 	}
 }
 
+func filterEmptyLayers(fsLayers []docker.FsLayer) (filteredLayers []docker.FsLayer) {
+	for _, layer := range fsLayers {
+		if layer.BlobSum != EMPTY_LAYER_BLOB_SUM {
+			filteredLayers = append(filteredLayers, layer)
+		}
+	}
+	return
+}
+
 // Analyse sent each layer from Docker image to Clair and returns
 // a list of found vulnerabilities
 func (c *Clair) Analyse(image *docker.Image) []Vulnerability {
+	// Filter the empty layers in image
+	image.FsLayers = filterEmptyLayers(image.FsLayers)
+	layerLength := len(image.FsLayers)
+	if layerLength == 0 {
+		fmt.Printf("No need to analyse image %s/%s:%s as there is no non-emtpy layer",
+			image.Registry, image.Name, image.Tag)
+		return nil
+	}
+
 	var vs []Vulnerability
-	for i := range image.FsLayers {
+	for i := layerLength - 1; i >= 0; i-- {
 		layer := newLayer(image, i)
 		err := c.pushLayer(layer)
 		if err != nil {
 			fmt.Printf("Push layer %d failed: %s", i, err.Error())
 			continue
 		}
-		lvs, err := c.analyzeLayer(layer)
-		if err != nil {
-			fmt.Printf("Analyze layer %d failed: %s", i, err.Error())
-		} else {
-			vs = append(vs, *lvs...)
-		}
 	}
+
+	vs, err := c.analyzeLayer(image.FsLayers[0])
+	if err != nil {
+		fmt.Printf("Analyse image %s/%s:%s failed: %s", image.Registry, image.Name, image.Tag, err.Error())
+		return nil
```

Comment:
The GET layer API will be called only once with the the top layer as it is out of loop. If it has error,  should return with no vulnerability.